### PR TITLE
Improved `Show` instance for String

### PR DESCRIPTION
**Pull request by fosskers**
Since there is no `Read` typeclass in Scalaz (nor should there be), `Show` becomes nearly lawless. That's okay if it's only used for debugging and never intended for serious serialization.

That said, here's an attempt at one law, which is reflected in the code change and alters the output of `.shows`:

> Law of String Identity: `String.shows == id`

### CODE REVIEW

File: core/src/main/scala/scalaz/std/String.scala [hunk 1]
```diff
@@ -6,7 +6,7 @@ trait StringInstances {
     type SA[α] = String
     def append(f1: String, f2: => String) = f1 + f2
     def zero: String = ""
-    override def show(f: String) = '"' + f + '"'
+    override def shows(f: String): String = f
```
Comment (fosskers) [inline core/src/main/scala/scalaz/std/String.scala]: Notice the change from `show` to `shows`.

Comment (edmundnoble) [reply core/src/main/scala/scalaz/std/String.scala]: I don't think it's a good idea to change behavior like this in 7.2.x. I do agree that this is how it should've been from the start, but that's beside the point.

**Comment by fosskers** [comment]
Lemme do a `publishLocal` and see. Right now we're looking at:
```
| Benchmark                               | ScalaZ 7.2.16 | ScalaZ 7.2.17 | Cats 1.0.0-RC1 | Vanilla Scala | Haskell 8.0.2 |
|-----------------------------------------+---------------+---------------+----------------+---------------+---------------|
| ~Show~ - ~String~                       | 216.6         |               | 3.2            | 2.8           | 199.4         |
```

**Comment by fosskers** [comment]
```
[info] ShowBench.showStrCats      avgt    5       3.308 ±     0.194  ns/op
[info] ShowBench.showStrScalaz    avgt    5       3.749 ±     0.104  ns/op
[info] ShowBench.showStrVanilla   avgt    5       2.877 ±     0.159  ns/op
```

**Comment by fosskers** [comment]
The only other discussion point is whether removing the surrounding quotes is okay. Haskell only adds them because its `Read` instance for `String` requires it.

**Comment by fommil** [comment]
definitely need the quotes around `String`. @edmundnoble I'm sure there's a way of making the change such that it's compatible with 7.2? Why not implement both methods?

**Comment by fosskers** [comment]
```
[error]  * method show(java.lang.String)scalaz.Cord in object scalaz.std.StringInstances#stringInstance's type is different in current version, where it is (java.lang.Object)scalaz.Cord instead of (java.lang.String)scalaz.Cord
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("scalaz.std.StringInstances#stringInstance.show")
java.lang.RuntimeException: scalaz-core: Binary compatibility check failed!
```
Zing, I'll add back an explicit `def show`.

**Comment by edmundnoble** [comment]
@fommil I'm pretty sure we're good now.

**Comment by fosskers** [comment]
Gross eh? That's another thing that threw me off.

**Comment by fommil** [comment]
I'd still have thought it'd be better to implement both methods independently. Fosskers you don't need to publish local to do perftests, just patch the monkey for great good.

**Comment by fommil** [comment]
Some numbers...


```
# scalaz 7.2.17
ShowBench.showStrScalaz  avgt   15  2167.086 ± 59.773  ns/op

# override def show(f: String): Cord = Cord.stringToCord(shows(f))
# override def shows(f: String): String = "\"" + f + "\""
[info] ShowBench.showStrScalaz  avgt   15  97.004 ± 1.336  ns/op

# override def show(f: String): Cord = new Cord(FingerTree.three("\"", f, "\"")(Cord.sizer).toTree)
# override def shows(f: String): String = "\"" + f + "\""
[info] ShowBench.showStrScalaz  avgt   15  96.869 ± 1.246  ns/op
```

this is the standard .show test in the https://github.com/fosskers/scalaz-and-cats

we can safely conclude that either approach is equivalent in terms of the cost to construct the Cord, but the latter has the best memory overhead so I think it is best. Consider the case where the strings are *huge*. I'll update the test to do that and also consider nested in a List.

**Comment by fommil** [comment]
I'm pretty amazed about the final result... doing `.toString` on a single `Cord` appears to do a *lot* more work than a no-op.

**Comment by fommil** [comment]
Needs a sync.

I forgot to check perf of this version with mixed char/string concatenation. Is it ok? My concern is that it's going via the anyToString implicit.

**Comment by fommil** [comment]
Oh, I'm also not sure if the fingertree is doing the right then, I never actually tested/printed it out. I assume it has an order preserving constructor.

**Comment by fommil** [comment]
yup, it's used correctly

```
scala> val cord = new Cord(FingerTree.three("\"", "hello", "\"")(Cord.sizer).toTree)
cord: scalaz.Cord = "hello"
```

**Pull request merged**